Fix exclude patterns '*' and '*:*'

eddsalkield opened the pull request:
Fix #256.

mike-burns commented:
If this fixes it, can you also update [the "bugs" section of lsrc.1](https://github.com/thoughtbot/rcm/blob/master/man/lsrc.1#L194-L199)?

Some prior art here is #11, #55, #58.

---

I pulled down your branch and added this test:

```
Keeps globs as globs

  $ mkdir vimulator
  > lsrc -vvv -x '*vim*' 2>&1 | grep exclude_file_globs
  exclude_file_globs: *vim*
```

it gave this failure:

```
FAIL: test/lsrc-undotted-star.t
===============================

!
--- test/lsrc-undotted-star.t
+++ test/lsrc-undotted-star.t.err
@@ -17,4 +17,4 @@
 
   $ mkdir vimulator
   > lsrc -vvv -x '*vim*' 2>&1 | grep exclude_file_globs
-  exclude_file_globs: *vim*
+  exclude_file_globs: trim

# Ran 1 tests, 0 skipped, 1 failed.
```

This might be fixed with more backticks. So I changed line 169 to

```
  for exclude in $(trim "$excludes"); do
```

and it gave me

```
FAIL: test/lsrc-undotted-star.t
===============================

!
--- test/lsrc-undotted-star.t
+++ test/lsrc-undotted-star.t.err
@@ -17,4 +17,4 @@
 
   $ mkdir vimulator
   > lsrc -vvv -x '*vim*' 2>&1 | grep exclude_file_globs
-  exclude_file_globs: *vim*
+  exclude_file_globs: vimulator

# Ran 1 tests, 0 skipped, 1 failed.
```

I don't totally understand what the `trim` function is trying to do or how it solves the `*` glob. Can you explain a bit?

eddsalkield commented:
There are several problems leading to the unintentional globbing issue:

Firstly, within `rcup` and `rcdn`, when constructing arguments to pass to `lsrc`, the _for_ loops over the arguments do not have quoted variables, leading to globbing.  I have quoted these accordingly.

Secondly, `lsrc` is invoked as follows:
`dests_and_srcs="$(lsrc $LS_ARGS)"`
When shells use command substitution like this, they go through two stages:
- Word expansion.  This is useful because it splits `LS_ARGS` back up into its constituent strings.
- File name expansion.  The side effect of this is to introduce globbing.

You can read more about how this works [here](https://www.tldp.org/LDP/Bash-Beginners-Guide/html/sect_03_04.html#sect_03_04_07).

To fix this, I have passed `lsrc` and its arguments to `eval`.  This involves quoting the relevant arguments, so:
```
  for dotfiles_dir in "$DOTFILES_DIRS"; do
    LS_ARGS="$LS_ARGS -d $dotfiles_dir"
  done
```
becomes
```
  for dotfiles_dir in "$DOTFILES_DIRS"; do
    LS_ARGS="$LS_ARGS -d \"$dotfiles_dir\""
  done
```
Then `lsrc` is invoked as follows:
`dests_and_srcs="$(eval "lsrc $LS_ARGS")"`

There is one final non-globbing issue: the parsing of arguments can introduce extra spaces in the variables, which then trip up the `dotfiles_dir_excludes` function.  For example:

`I) includes="$includes $OPTARG";;`

introduces a space if `includes` is empty or null.

This is what `trim` was supposed to address, by removing the whitespace.  I have now replaced this hacky solution by instead introducing a function `append_variable`, which allows two variables to be appended without introducing unnecessary whitespace.  Then the additional whitespace is never added in the first place.

I have updated my branch with your test, which now passes, alongside the requested update to the documentation.

mike-burns commented:
That looks like it did it! Amazing.

I have merged this as cbc346b279b5c08b281c9901b994a58eb037a60d. I borrowed your great explanation for the commit message.

I haven't yet tried this on additional platforms -- only Debian dash for now. I'd like to do that and then cut a new release.

Thank you again for thinking through this!
